### PR TITLE
Allow runtime selection of models with `:CopilotChatModels`

### LINKED ISSUE #366: feat: gpt-4o support
Hi,
copilot chat now supports gpt-4o. Could you add its support? - I think it would need to use https://openaipublic.blob.core.windows.net/encodings/o200k_base.tiktoken

Thanks.

### PR DESCRIPTION
**Pull request by gptlang**
CC @deathbeam & CC @jellydn 

I haven't touched Lua in a while and don't really know how the codebase works anymore. Could you guys check if I did anything wrong

Tentatively closes #366

### CODE REVIEW

**Comment by pidgeon777** [comment]
So, to clarify, it will be possible to dynamically choose which model to use for any given request?

**Comment by gptlang** [comment]
Well, you'll have to specify it. It's not automagic like vscode. It's currently a bit slow due to fetching the models on each request. We might want to cache it per session.

**Comment by gptlang** [comment]
But the point is to make discovery of available models easier, for example to check if gpt-4o is available to you

**Comment by jellydn** [comment]
Hi, @gptlang What if we save the selected model and use that config for next time? It's similar to what `Copilot Auth` does. So we don't need to call models every time. What do you think?

**Comment by gptlang** [comment]
I don't want to touch `init.lua` and would rather have that be followed by default. If you want to use a model by default, set it there

**Review by jellydn — APPROVED** [review]
LGTM!

File: lua/CopilotChat/init.lua [hunk 1]
```diff
@@ -354,11 +354,25 @@ function M.toggle(config, source)
...
 end
 
--- @returns string
+--- @returns string
 function M.response()
   return state.response
 end
 
+function M.select_model()
+  state.copilot:select_model(function(model)
```
Comment (deathbeam) [inline lua/CopilotChat/init.lua:363]: this should only change M.config, and the config itself shouldnt ever be nil, it shouldnt change stateconfig.model or default_config.model

File: lua/CopilotChat/copilot.lua [hunk 2]
```diff
@@ -499,6 +501,55 @@ function Copilot:ask(prompt, opts)
   end, on_error)
 end
 
+--- Fetch & allow model selection
+---@param callback fun(string):nil
+function Copilot:select_model(callback)
+  if self.models_cache ~= nil then
```
Comment (deathbeam) [inline lua/CopilotChat/copilot.lua:507]: you can reduce the code duplication here by making inner local function that does the vim.ui.select. but also i think the select shouldnt be here at all, it should be in the select_model method in init.lua, this should simply call the callback with the model_cache

**Review by deathbeam — CHANGES_REQUESTED** [review]

**Pull request merged**